### PR TITLE
Lazy objects v0 for raw objects

### PR DESCRIPTION
**Pull request by yinghuitan**
Release Note: introduce lazy objects feature for raw objects in Prepack.
Introduce lazy objects feature for raw objects in Prepack with supported runtime.
Lazy object is a collaboration feature between the Prepack and new runtime that supports this feature. With this feature, object can be created with a special LazyObjectsRuntime.createLazyObject() API. The object created with this API can be a light-weight placeholder object. Once the lazy object is *touched/used* the first time, runtime is responsible to callback the registered initialization function to hydrate the object into a real object. 
By doing this we will save both memory and object creation speed. 

Currently, Prepack only supports lazy for raw objects.

The APIs that runtime needs to support:
1. LazyObjectsRuntime.createLazyObject(<lazy_object_id>);
2. LazyObjectsRuntime.setLazyObjectInitializer(<well-known-callback>).

Under lazy objects mode, Prepack will serialize all lazy objects into two parts:
 1. All lazy objects are created via lightweight LazyObjectsRuntime.createLazyObject() call.
 2. Lazy objects' property assignments are delayed in a callback function which is registered with the runtime. 
 
TODO:
1. Test-runner support for lazy mode.
2. Support more object types for this feature
3. Remove "reference count === 1" objects' root and directly serialize into callback function.
4. Figure out "reference count > 1" objects that are safe to remove root from global anonymous function and directly serialize into callback function.

Test Plan:
1. "bin/prepack.js <foo.js> --lazyObjectsRuntime LazyObjectRuntime"
2. test-runner introduced a lazy mode(later PR) can successfully run all tests.

### CODE REVIEW

File: src/prepack-options.js [hunk 1]
```diff
@@ -101,6 +105,9 @@ export function getSerializerOptions({
     trace,
   };
   if (additionalFunctions) result.additionalFunctions = additionalFunctions;
+  if (lazyObjectsRuntime != null) {
```
Comment (NTillmann) [inline src/prepack-options.js]: I'd prefer !== undefined

File: src/serializer/ResidualHeapSerializer.js [hunk 2]
```diff
@@ -1202,7 +1212,6 @@ export class ResidualHeapSerializer {
   }
 
   _serializeAbstractValueHelper(val: AbstractValue): BabelNodeExpression {
-    invariant(val.kind !== "sentinel member expression", "invariant established by visitor");
```
Comment (NTillmann) [inline src/serializer/ResidualHeapSerializer.js]: Why did you remove this invariant?

File: src/serializer/LazyObjectsSerializer.js [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    this._initializationCallbackName = t.identifier("__initializerCallback");
+    this._currentSerializeLazyObject = null;
+  }
+
+  _lazyObjectIdSeed: number;
+  _valueLazyIds: Map<ObjectValue, number>;
+  // Holds object's lazy initializer bodies.
+  // These bodies will be combined into a well-known callback after generator serialization is done and registered with the runtime.
+  _lazyObjectInitializers: Map<ObjectValue, SerializedBody>;
+  _currentSerializeLazyObject: ?ObjectValue;
```
Comment (NTillmann) [inline src/serializer/LazyObjectsSerializer.js]: void | ObjectValue. No need to allow two kinds of undefinedness.

File: src/serializer/ResidualHeapSerializer.js (hunk 2 above)
Comment (yinghuitan) [reply src/serializer/ResidualHeapSerializer.js]: This is a redundant invariant from previous checkin.
In "_serializeAbstractValue()" function, there is already same invariant.
